File: packages/web/src/lib/expressionEvaluator.ts
/**
 * Family-formula expression evaluator — FAM-04.
 *
 * Safe recursive-descent parser. Supports the §13.3 numeric grammar
 * plus the conditional / boolean grammar Revit uses for parameter
 * formulas. Booleans are encoded as 1.0 / 0.0 (Revit-compatible).
 *
 * Supported constructs:
 *   - Arithmetic: + - * / mod ** ()
 *   - Comparison: < <= > >= = <>  (return 1.0 or 0.0)
 *   - Functions:  if(cond, a, b), rounddown, roundup, round,
 *                 min, max, abs, sqrt, mod, not, and, or
 *   - Identifiers: bare names referencing the supplied parameter map.
 *
 * Unknown identifiers raise. No global access — all symbols pass
 * through the whitelisted function table or the explicit parameter
 * map. Never use eval / new Function on the input.
 */

type Token =
  | { kind: 'num'; value: number }
  | { kind: 'ident'; value: string }
  | { kind: 'op'; value: string }
  | { kind: 'lparen' }
  | { kind: 'rparen' }
  | { kind: 'comma' }
  | { kind: 'eof' };

const SINGLE_OPS = new Set(['+', '-', '*', '/', '%', '=']);

function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  while (i < input.length) {
    const ch = input[i];
    if (ch === ' ' || ch === '\t' || ch === '\n' || ch === '\r') {
      i++;
      continue;
    }
    if (ch === '(') {
      tokens.push({ kind: 'lparen' });
      i++;
      continue;
    }
    if (ch === ')') {
      tokens.push({ kind: 'rparen' });
      i++;
      continue;
    }
    if (ch === ',') {
      tokens.push({ kind: 'comma' });
      i++;
      continue;
    }
    // Multi-char operators first
    if (input.startsWith('**', i)) {
      tokens.push({ kind: 'op', value: '**' });
      i += 2;
      continue;
    }
    if (input.startsWith('<=', i)) {
      tokens.push({ kind: 'op', value: '<=' });
      i += 2;
      continue;
    }
    if (input.startsWith('>=', i)) {
      tokens.push({ kind: 'op', value: '>=' });
      i += 2;
      continue;
    }
    if (input.startsWith('<>', i)) {
      tokens.push({ kind: 'op', value: '<>' });
      i += 2;
      continue;
    }
    if (ch === '<') {
      tokens.push({ kind: 'op', value: '<' });
      i++;
      continue;
    }
    if (ch === '>') {
      tokens.push({ kind: 'op', value: '>' });
      i++;
      continue;
    }
    if (SINGLE_OPS.has(ch)) {
      tokens.push({ kind: 'op', value: ch });
      i++;
      continue;
    }
    if ((ch >= '0' && ch <= '9') || ch === '.') {
      let j = i;
      let dot = ch === '.';
      while (
        j + 1 < input.length &&
        ((input[j + 1] >= '0' && input[j + 1] <= '9') || (input[j + 1] === '.' && !dot))
      ) {
        j++;
        if (input[j] === '.') dot = true;
      }
      const numStr = input.slice(i, j + 1);
      const value = Number(numStr);
      if (!Number.isFinite(value)) {
        throw new Error(`expressionEvaluator: invalid number "${numStr}"`);
      }
      tokens.push({ kind: 'num', value });
      i = j + 1;
      continue;
    }
    if ((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || ch === '_') {
      let j = i;
      while (
        j + 1 < input.length &&
        ((input[j + 1] >= 'a' && input[j + 1] <= 'z') ||
          (input[j + 1] >= 'A' && input[j + 1] <= 'Z') ||
          (input[j + 1] >= '0' && input[j + 1] <= '9') ||
          input[j + 1] === '_')
      ) {
        j++;
      }
      tokens.push({ kind: 'ident', value: input.slice(i, j + 1) });
      i = j + 1;
      continue;
    }
    throw new Error(`expressionEvaluator: unexpected character "${ch}" at index ${i}`);
  }
  tokens.push({ kind: 'eof' });
  return tokens;
}

class Parser {
  private pos = 0;

  constructor(
    private tokens: Token[],
    private params: Record<string, number | boolean>,
  ) {}

  evaluate(): number {
    const value = this.parseComparison();
    if (this.peek().kind !== 'eof') {
      throw new Error('expressionEvaluator: unexpected trailing input');
    }
    return value;
  }

  private peek(): Token {
    return this.tokens[this.pos];
  }

  private consume(): Token {
    return this.tokens[this.pos++];
  }

  private expect(predicate: (t: Token) => boolean, msg: string): Token {
    const t = this.consume();
    if (!predicate(t)) throw new Error(`expressionEvaluator: ${msg}`);
    return t;
  }

  private parseComparison(): number {
    let left = this.parseAdd();
    const t = this.peek();
    if (t.kind === 'op' && ['<', '<=', '>', '>=', '=', '<>'].includes(t.value)) {
      const op = t.value;
      this.consume();
      const right = this.parseAdd();
      switch (op) {
        case '<':
          return left < right ? 1 : 0;
        case '<=':
          return left <= right ? 1 : 0;
        case '>':
          return left > right ? 1 : 0;
        case '>=':
          return left >= right ? 1 : 0;
        case '=':
          return left === right ? 1 : 0;
        case '<>':
          return left !== right ? 1 : 0;
      }
    }
    return left;
  }

  private parseAdd(): number {
    let left = this.parseMul();
    while (true) {
      const t = this.peek();
      if (t.kind === 'op' && (t.value === '+' || t.value === '-')) {
        this.consume();
        const right = this.parseMul();
        left = t.value === '+' ? left + right : left - right;
      } else {
        return left;
      }
    }
  }

  private parseMul(): number {
    let left = this.parsePow();
    while (true) {
      const t = this.peek();
      if (t.kind === 'op' && (t.value === '*' || t.value === '/' || t.value === '%')) {
        this.consume();
        const right = this.parsePow();
        if (t.value === '*') left = left * right;
        else if (t.value === '/') left = left / right;
        else left = left - Math.trunc(left / right) * right; // mod
      } else if (t.kind === 'ident' && t.value.toLowerCase() === 'mod') {
        // bare `mod` infix (Revit spelling)
        this.consume();
        const right = this.parsePow();
        left = left - Math.trunc(left / right) * right;
      } else {
        return left;
      }
    }
  }

  private parsePow(): number {
    const left = this.parseUnary();
    const t = this.peek();
    if (t.kind === 'op' && t.value === '**') {
      this.consume();
      const right = this.parsePow(); // right-assoc
      return Math.pow(left, right);
    }
    return left;
  }

  private parseUnary(): number {
    const t = this.peek();
    if (t.kind === 'op' && t.value === '-') {
      this.consume();
      return -this.parseUnary();
    }
    if (t.kind === 'op' && t.value === '+') {
      this.consume();
      return this.parseUnary();
    }
    return this.parsePrimary();
  }

  private parsePrimary(): number {
    const t = this.consume();
    if (t.kind === 'num') return t.value;
    if (t.kind === 'lparen') {
      const v = this.parseComparison();
      this.expect((x) => x.kind === 'rparen', 'expected ")"');
      return v;
    }
    if (t.kind === 'ident') {
      const name = t.value;
      if (this.peek().kind === 'lparen') {
        this.consume(); // (
        const args: number[] = [];
        if (this.peek().kind !== 'rparen') {
          args.push(this.parseComparison());
          while (this.peek().kind === 'comma') {
            this.consume();
            args.push(this.parseComparison());
          }
        }
        this.expect((x) => x.kind === 'rparen', 'expected ")" closing function call');
        return callFunction(name, args);
      }
      // Identifier reference: parameters table or constant true/false.
      const lower = name.toLowerCase();
      if (lower === 'true') return 1;
      if (lower === 'false') return 0;
      if (lower === 'pi') return Math.PI;
      if (lower === 'e') return Math.E;
      if (!(name in this.params)) {
        throw new Error(`expressionEvaluator: unknown identifier "${name}"`);
      }
      const v = this.params[name];
      return typeof v === 'boolean' ? (v ? 1 : 0) : Number(v);
    }
    throw new Error(`expressionEvaluator: unexpected token ${JSON.stringify(t)}`);
  }
}

function callFunction(rawName: string, args: number[]): number {
  const name = rawName.toLowerCase();
  switch (name) {
    case 'if': {
      if (args.length !== 3) throw new Error('if(cond, a, b) needs 3 args');
      return args[0] !== 0 ? args[1] : args[2];
    }
    case 'rounddown': {
      if (args.length !== 1) throw new Error('rounddown(x) needs 1 arg');
      return Math.floor(args[0]);
    }
    case 'roundup': {
      if (args.length !== 1) throw new Error('roundup(x) needs 1 arg');
      return Math.ceil(args[0]);
    }
    case 'round': {
      if (args.length !== 1) throw new Error('round(x) needs 1 arg');
      return Math.round(args[0]);
    }
    case 'min': {
      if (args.length === 0) throw new Error('min needs at least 1 arg');
      return Math.min(...args);
    }
    case 'max': {
      if (args.length === 0) throw new Error('max needs at least 1 arg');
      return Math.max(...args);
    }
    case 'abs': {
      if (args.length !== 1) throw new Error('abs(x) needs 1 arg');
      return Math.abs(args[0]);
    }
    case 'sqrt': {
      if (args.length !== 1) throw new Error('sqrt(x) needs 1 arg');
      return Math.sqrt(args[0]);
    }
    case 'mod': {
      if (args.length !== 2) throw new Error('mod(a,b) needs 2 args');
      return args[0] - Math.trunc(args[0] / args[1]) * args[1];
    }
    case 'not': {
      if (args.length !== 1) throw new Error('not(x) needs 1 arg');
      return args[0] === 0 ? 1 : 0;
    }
    case 'and': {
      if (args.length < 2) throw new Error('and needs at least 2 args');
      return args.every((a) => a !== 0) ? 1 : 0;
    }
    case 'or': {
      if (args.length < 2) throw new Error('or needs at least 2 args');
      return args.some((a) => a !== 0) ? 1 : 0;
    }
    default:
      throw new Error(`expressionEvaluator: unknown function "${rawName}"`);
  }
}

/**
 * Evaluate a family-formula expression with the supplied parameter map.
 *
 * Returns `null` on parse error or when the result is not a finite
 * number — callers surface the failure inline rather than throwing.
 */
export function evaluateFormula(
  raw: string,
  params: Record<string, number | boolean> = {},
): number | null {
  const trimmed = raw.trim();
  if (!trimmed) return null;
  try {
    const tokens = tokenize(trimmed);
    const parser = new Parser(tokens, params);
    const value = parser.evaluate();
    return Number.isFinite(value) ? value : null;
  } catch {
    return null;
  }
}

/** Same as :func:`evaluateFormula` but throws the parser error so the
 *  family editor can show "expected `)` at column 12"-style messages. */
export function evaluateFormulaOrThrow(
  raw: string,
  params: Record<string, number | boolean> = {},
): number {
  const trimmed = raw.trim();
  if (!trimmed) throw new Error('empty expression');
  const tokens = tokenize(trimmed);
  const parser = new Parser(tokens, params);
  const value = parser.evaluate();
  if (!Number.isFinite(value)) throw new Error('non-finite result');
  return value;
}

/** Lightweight pre-flight: returns null on success or the error message. */
export function validateFormula(raw: string, knownParams: Iterable<string> = []): string | null {
  const trimmed = raw.trim();
  if (!trimmed) return null;
  const stub: Record<string, number> = {};
  for (const k of knownParams) stub[k] = 0;
  try {
    const tokens = tokenize(trimmed);
    new Parser(tokens, stub).evaluate();
    return null;
  } catch (e) {
    return e instanceof Error ? e.message : String(e);
  }
}

File: app/bim_ai/expression_evaluator.py
"""Family-formula expression evaluator — FAM-04 (Python side).

Mirror of ``packages/web/src/lib/expressionEvaluator.ts``. Safe
recursive-descent parser; never uses ``eval`` / ``exec``. Booleans
are encoded as 1.0 / 0.0 (Revit-compatible).

Supported constructs:
    - Arithmetic:  + - * / ** ()  and ``mod`` / ``%`` for modulo
    - Comparison:  < <= > >= = <>   (return 1.0 or 0.0)
    - Functions:   if(cond, a, b), rounddown, roundup, round,
                   min, max, abs, sqrt, mod, not, and, or
    - Identifiers: bare names referencing the supplied parameter map.

Unknown identifiers raise ``ExpressionError``. No global access.
"""

from __future__ import annotations

import math
from collections.abc import Iterable, Mapping
from dataclasses import dataclass


class ExpressionError(ValueError):
    """Raised on parse or evaluation failure."""


_SINGLE_OPS = {"+", "-", "*", "/", "%", "="}


@dataclass(frozen=True)
class _Token:
    kind: str  # 'num', 'ident', 'op', 'lparen', 'rparen', 'comma', 'eof'
    value: str | float | None = None


def _tokenize(source: str) -> list[_Token]:
    tokens: list[_Token] = []
    i = 0
    n = len(source)
    while i < n:
        ch = source[i]
        if ch.isspace():
            i += 1
            continue
        if ch == "(":
            tokens.append(_Token("lparen"))
            i += 1
            continue
        if ch == ")":
            tokens.append(_Token("rparen"))
            i += 1
            continue
        if ch == ",":
            tokens.append(_Token("comma"))
            i += 1
            continue
        if source.startswith("**", i):
            tokens.append(_Token("op", "**"))
            i += 2
            continue
        if source.startswith("<=", i):
            tokens.append(_Token("op", "<="))
            i += 2
            continue
        if source.startswith(">=", i):
            tokens.append(_Token("op", ">="))
            i += 2
            continue
        if source.startswith("<>", i):
            tokens.append(_Token("op", "<>"))
            i += 2
            continue
        if ch in "<>":
            tokens.append(_Token("op", ch))
            i += 1
            continue
        if ch in _SINGLE_OPS:
            tokens.append(_Token("op", ch))
            i += 1
            continue
        if ch.isdigit() or ch == ".":
            j = i
            saw_dot = ch == "."
            while j + 1 < n and (source[j + 1].isdigit() or (source[j + 1] == "." and not saw_dot)):
                j += 1
                if source[j] == ".":
                    saw_dot = True
            num_str = source[i : j + 1]
            try:
                value = float(num_str)
            except ValueError as exc:
                raise ExpressionError(f"invalid number {num_str!r}") from exc
            if not math.isfinite(value):
                raise ExpressionError(f"non-finite number {num_str!r}")
            tokens.append(_Token("num", value))
            i = j + 1
            continue
        if ch.isalpha() or ch == "_":
            j = i
            while j + 1 < n and (source[j + 1].isalnum() or source[j + 1] == "_"):
                j += 1
            tokens.append(_Token("ident", source[i : j + 1]))
            i = j + 1
            continue
        raise ExpressionError(f"unexpected character {ch!r} at index {i}")
    tokens.append(_Token("eof"))
    return tokens


def _is_truthy(x: float) -> bool:
    return x != 0


def _call_function(name: str, args: list[float]) -> float:
    n = name.lower()
    if n == "if":
        if len(args) != 3:
            raise ExpressionError("if(cond, a, b) needs 3 args")
        return args[1] if _is_truthy(args[0]) else args[2]
    if n == "rounddown":
        if len(args) != 1:
            raise ExpressionError("rounddown(x) needs 1 arg")
        return float(math.floor(args[0]))
    if n == "roundup":
        if len(args) != 1:
            raise ExpressionError("roundup(x) needs 1 arg")
        return float(math.ceil(args[0]))
    if n == "round":
        if len(args) != 1:
            raise ExpressionError("round(x) needs 1 arg")
        # Python's round uses banker's rounding; Revit/JS use half-away-from-zero.
        x = args[0]
        return float(math.floor(x + 0.5)) if x >= 0 else -float(math.floor(-x + 0.5))
    if n == "min":
        if not args:
            raise ExpressionError("min needs at least 1 arg")
        return float(min(args))
    if n == "max":
        if not args:
            raise ExpressionError("max needs at least 1 arg")
        return float(max(args))
    if n == "abs":
        if len(args) != 1:
            raise ExpressionError("abs(x) needs 1 arg")
        return float(abs(args[0]))
    if n == "sqrt":
        if len(args) != 1:
            raise ExpressionError("sqrt(x) needs 1 arg")
        return float(math.sqrt(args[0]))
    if n == "mod":
        if len(args) != 2:
            raise ExpressionError("mod(a, b) needs 2 args")
        a, b = args
        return a - math.trunc(a / b) * b
    if n == "not":
        if len(args) != 1:
            raise ExpressionError("not(x) needs 1 arg")
        return 0.0 if _is_truthy(args[0]) else 1.0
    if n == "and":
        if len(args) < 2:
            raise ExpressionError("and needs at least 2 args")
        return 1.0 if all(_is_truthy(a) for a in args) else 0.0
    if n == "or":
        if len(args) < 2:
            raise ExpressionError("or needs at least 2 args")
        return 1.0 if any(_is_truthy(a) for a in args) else 0.0
    raise ExpressionError(f"unknown function {name!r}")


class _Parser:
    def __init__(self, tokens: list[_Token], params: Mapping[str, float | bool]) -> None:
        self._tokens = tokens
        self._params = params
        self._pos = 0

    def evaluate(self) -> float:
        value = self._comparison()
        if self._peek().kind != "eof":
            raise ExpressionError("unexpected trailing input")
        return value

    def _peek(self) -> _Token:
        return self._tokens[self._pos]

    def _consume(self) -> _Token:
        t = self._tokens[self._pos]
        self._pos += 1
        return t

    def _expect_kind(self, kind: str, msg: str) -> _Token:
        t = self._consume()
        if t.kind != kind:
            raise ExpressionError(msg)
        return t

    def _comparison(self) -> float:
        left = self._add()
        t = self._peek()
        if t.kind == "op" and t.value in ("<", "<=", ">", ">=", "=", "<>"):
            op = t.value
            self._consume()
            right = self._add()
            if op == "<":
                return 1.0 if left < right else 0.0
            if op == "<=":
                return 1.0 if left <= right else 0.0
            if op == ">":
                return 1.0 if left > right else 0.0
            if op == ">=":
                return 1.0 if left >= right else 0.0
            if op == "=":
                return 1.0 if left == right else 0.0
            if op == "<>":
                return 1.0 if left != right else 0.0
        return left

    def _add(self) -> float:
        left = self._mul()
        while True:
            t = self._peek()
            if t.kind == "op" and t.value in ("+", "-"):
                self._consume()
                right = self._mul()
                left = left + right if t.value == "+" else left - right
            else:
                return left

    def _mul(self) -> float:
        left = self._pow()
        while True:
            t = self._peek()
            if t.kind == "op" and t.value in ("*", "/", "%"):
                self._consume()
                right = self._pow()
                if t.value == "*":
                    left = left * right
                elif t.value == "/":
                    left = left / right
                else:
                    left = left - math.trunc(left / right) * right
            elif t.kind == "ident" and isinstance(t.value, str) and t.value.lower() == "mod":
                self._consume()
                right = self._pow()
                left = left - math.trunc(left / right) * right
            else:
                return left

    def _pow(self) -> float:
        left = self._unary()
        t = self._peek()
        if t.kind == "op" and t.value == "**":
            self._consume()
            right = self._pow()  # right-assoc
            return float(math.pow(left, right))
        return left

    def _unary(self) -> float:
        t = self._peek()
        if t.kind == "op" and t.value == "-":
            self._consume()
            return -self._unary()
        if t.kind == "op" and t.value == "+":
            self._consume()
            return self._unary()
        return self._primary()

    def _primary(self) -> float:
        t = self._consume()
        if t.kind == "num":
            assert isinstance(t.value, float)
            return t.value
        if t.kind == "lparen":
            v = self._comparison()
            self._expect_kind("rparen", 'expected ")"')
            return v
        if t.kind == "ident":
            assert isinstance(t.value, str)
            name = t.value
            if self._peek().kind == "lparen":
                self._consume()
                args: list[float] = []
                if self._peek().kind != "rparen":
                    args.append(self._comparison())
                    while self._peek().kind == "comma":
                        self._consume()
                        args.append(self._comparison())
                self._expect_kind("rparen", 'expected ")" closing function call')
                return _call_function(name, args)
            lower = name.lower()
            if lower == "true":
                return 1.0
            if lower == "false":
                return 0.0
            if lower == "pi":
                return math.pi
            if lower == "e":
                return math.e
            if name not in self._params:
                raise ExpressionError(f"unknown identifier {name!r}")
            v = self._params[name]
            if isinstance(v, bool):
                return 1.0 if v else 0.0
            return float(v)
        raise ExpressionError(f"unexpected token {t!r}")


def evaluate_formula(
    raw: str, params: Mapping[str, float | bool] | None = None
) -> float | None:
    """Evaluate a family-formula expression. Returns ``None`` on failure."""
    trimmed = raw.strip()
    if not trimmed:
        return None
    try:
        tokens = _tokenize(trimmed)
        parser = _Parser(tokens, params or {})
        value = parser.evaluate()
        return value if math.isfinite(value) else None
    except ExpressionError:
        return None


def evaluate_formula_or_throw(
    raw: str, params: Mapping[str, float | bool] | None = None
) -> float:
    """Same as :func:`evaluate_formula` but raises ``ExpressionError``."""
    trimmed = raw.strip()
    if not trimmed:
        raise ExpressionError("empty expression")
    tokens = _tokenize(trimmed)
    parser = _Parser(tokens, params or {})
    value = parser.evaluate()
    if not math.isfinite(value):
        raise ExpressionError("non-finite result")
    return value


def validate_formula(raw: str, known_params: Iterable[str] = ()) -> str | None:
    """Pre-flight: returns ``None`` on success or the error message."""
    trimmed = raw.strip()
    if not trimmed:
        return None
    stub: dict[str, float] = {k: 0.0 for k in known_params}
    try:
        tokens = _tokenize(trimmed)
        _Parser(tokens, stub).evaluate()
        return None
    except ExpressionError as exc:
        return str(exc)
